lib/Target/RISCV/RISCVHelper.h

```diff
@@ -210,13 +210,15 @@ template <typename T> T getPCRELLO(T Value, T PC) {
 /// The following instruction types are supported
 /// I, S, SB, U, UJ, RVC, RVC(LUI), RVC(B), RVC(J)
 ///------------------------------------------
-template <typename T> T encodeI(T Result) {
-  Result &= 0xFFF;
-  Result <<= 20;
-  return Result;
+inline int32_t encodeI(int32_t Result) {
+  int32_t Hi20 = (Result + 0x800) & 0xfffff000;
+  Result -= Hi20;
+  return (Result & 0xfff) << 20;
 }
 
-template <typename T> T encodeS(T &Result) {
+inline int32_t encodeS(int32_t Result) {
+  int32_t Hi20 = (Result + 0x800) & 0xfffff000;
+  Result -= Hi20;
   uint32_t Imm11_5 = extractBits(Result, 11, 5) << 25;
   uint32_t Imm4_0 = extractBits(Result, 4, 0) << 7;
   Result = Imm11_5 | Imm4_0;
```

Comment:
Again, this seems more complex than necessary. You should just be able to extract the bits, from `Result` and insert them into the right place of the result